src/data/list/rotate.lean

```diff
@@ -102,7 +111,236 @@ lemma prod_rotate_eq_one_of_prod_eq_one [group α] : ∀ {l : list α} (hl : l.p
+  rw [rotate_eq_drop_append_take_mod, rotate_eq_drop_append_take_mod] at h,
+  obtain ⟨hd, ht⟩ := append_inj h _,
+  { rw [←take_append_drop _ l, ht, hd, take_append_drop] },
+  { rw [length_drop, length_drop, hle] }
+end
+
+-- possibly easier to find in doc-gen, otherwise not that useful.
+lemma rotate_eq_rotate {l l' : list α} {n : ℕ} :
+  l.rotate n = l'.rotate n ↔ l = l' :=
+(rotate_injective n).eq_iff
+
+lemma rotate_eq_iff {l l' : list α} {n : ℕ} :
+  l.rotate n = l' ↔ l = l'.rotate (l'.length - n % l'.length) :=
+begin
+  rw [←@rotate_eq_rotate _ l _ n, rotate_rotate, ←rotate_mod l', add_mod],
+  cases l'.length.zero_le.eq_or_lt with hl hl,
+  { rw [eq_nil_of_length_eq_zero hl.symm, rotate_nil, rotate_eq_nil_iff] },
+  { cases (nat.zero_le (n % l'.length)).eq_or_lt with hn hn,
+    { simp [←hn] },
+    { rw [mod_eq_of_lt (nat.sub_lt_self hl hn), nat.sub_add_cancel, mod_self, rotate_zero],
+      exact (nat.mod_lt _ hl).le } }
+end
+
+lemma reverse_rotate (l : list α) (n : ℕ) :
+  (l.rotate n).reverse = l.reverse.rotate (l.length - (n % l.length)) :=
+begin
+  rw [←length_reverse l, ←rotate_eq_iff],
+  induction n with n hn generalizing l,
+  { simp },
+  { cases l with hd tl,
+    { simp },
+    { rw [rotate_cons_succ, nat.succ_eq_add_one, ←rotate_rotate, hn],
+      simp } }
+end
+
+section is_rotated
+
+variables (l l' : list α)
+
+/-- `is_rotated l₁ l₂` or `l₁ ~r l₂` asserts that `l₁` and `l₂` are cyclic permutations
+  of each other. This is defined by claiming that `∃ n, l.rotate n = l'`. -/
+def is_rotated : Prop := ∃ n, l.rotate n = l'
+
+infixr ` ~r `:1000 := is_rotated
+
+variables {l l'}
+
+@[refl] lemma is_rotated.refl (l : list α) : l ~r l :=
+⟨0, by simp⟩
+
+@[symm] lemma is_rotated.symm (h : l ~r l') : l' ~r l :=
+begin
+  obtain ⟨n, rfl⟩ := h,
+  cases l with hd tl,
+  { simp },
+  { use (hd :: tl).length * n - n,
+    rw [rotate_rotate, nat.add_sub_cancel', rotate_length_mul],
+    exact nat.le_mul_of_pos_left (by simp) }
+end
+
+lemma is_rotated_comm : l ~r l' ↔ l' ~r l :=
+⟨is_rotated.symm, is_rotated.symm⟩
+
+@[simp] protected lemma is_rotated.forall (l : list α) (n : ℕ) : l.rotate n ~r l :=
+is_rotated.symm ⟨n, rfl⟩
+
+@[trans] lemma is_rotated.trans {l'' : list α} (h : l ~r l') (h' : l' ~r l'') :
+  l ~r l'' :=
+begin
+  obtain ⟨n, rfl⟩ := h,
+  obtain ⟨m, rfl⟩ := h',
+  rw rotate_rotate,
+  use (n + m)
+end
+
+lemma is_rotated.eqv : equivalence (@is_rotated α) :=
+mk_equivalence _ is_rotated.refl (λ _ _, is_rotated.symm) (λ _ _ _, is_rotated.trans)
+
+/-- The relation `list.is_rotated l l'` forms a `setoid` of cycles. -/
+def is_rotated.setoid (α : Type*) : setoid (list α) :=
+{ r := is_rotated, iseqv := is_rotated.eqv }
+
+lemma is_rotated.perm (h : l ~r l') : l ~ l' :=
+exists.elim h (λ _ hl, hl ▸ (rotate_perm _ _).symm)
+
+lemma is_rotated.nodup_iff (h : l ~r l') : nodup l ↔ nodup l' :=
+h.perm.nodup_iff
+
+lemma nodup.of_is_rotated (h : nodup l) (h' : l ~r l') : nodup l' :=
+h'.nodup_iff.mp h
```

Comment:
Does this API exist for list.perm? I think we should either add it for just perm, both perm and is_rotated, or not add it at all.

Reply:
Removed.